### PR TITLE
Feature refactor apply

### PR DESCRIPTION
**Pull request by sopel39**
Fixes https://github.com/prestodb/presto/issues/6637

FYI: @martint

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 1]
```diff
@@ -37,22 +39,30 @@
      */
     private final List<Symbol> correlation;
 
+    /**
+     * Non-scalar predicates which operate on subqueries (e.g: "expr IN (SELECT 1)")
+     */
+    private final Map<Symbol, Expression> subqueryPredicates;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: why do you limit that only to non scalar? I would store all the subquery expressions

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 2]
```diff
@@ -87,6 +103,7 @@ public PlanNode getSubquery()
         return ImmutableList.<Symbol>builder()
                 .addAll(input.getOutputSymbols())
                 .addAll(subquery.getOutputSymbols())
+                .addAll(subqueryPredicates.keySet())
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: I think you should not add `subquery.getOutputSymbols()` when you add subqeryPredicates

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: I would name it `subqueryAssignment`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java [hunk 3]
```diff
@@ -74,197 +61,52 @@
...
+            Symbol semiJoinSymbol = getOnlyElement(node.getSubqueryPredicates().keySet());
+            return new SemiJoinNode(idAllocator.getNextId(),
+                    input,
+                    subquery,
+                    Symbol.from(inPredicate.getValue()),
+                    Symbol.from(inPredicate.getValueList()),
+                    semiJoinSymbol,
+                    Optional.empty(),
+                    Optional.empty()
+            );
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: nice!! :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: maybe add some javadoc to `ApplyNode` class

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 4]
```diff
@@ -637,6 +601,44 @@ private void map(Symbol symbol, Symbol canonical)
             mapping.put(symbol, canonical);
         }
 
+        private Map<Symbol, Expression> canonicalize(Map<Symbol, Expression> oldAssignments)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java]: extract this as separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java [hunk 5]
```diff
@@ -53,7 +52,7 @@ public Rewriter(PlanNodeIdAllocator idAllocator)
         public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
         {
             ApplyNode rewrittenNode = (ApplyNode) context.defaultRewrite(node, context.get());
-            if (rewrittenNode.getCorrelation().isEmpty() && rewrittenNode.getSubquery() instanceof EnforceSingleRowNode) {
+            if (rewrittenNode.getCorrelation().isEmpty() && rewrittenNode.getSubqueryPredicates().isEmpty()) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: I would undo this

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: maybe use `ScalarQueryUtil`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java (hunk 4 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java]: I don't think it is possible to extract those changes to a separate commit. The reason is that since subquery is unaliased, the assignments have to be unaliased too. Otherwise there will be unreferenced symbols and validator would fail

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: I will store subquery expressions in separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java (hunk 5 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: renamed to `subqueryAssignments`

**Comment by sopel39** [comment]
refactored & fixed q20.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 6]
```diff
@@ -206,14 +206,18 @@ private PlanBuilder appendInPredicateApplyNode(PlanBuilder subPlan, InPredicate
 
         TranslationMap translationMap = subPlan.copyTranslations();
         InPredicate parametersReplaced = ExpressionTreeRewriter.rewriteWith(new ParameterRewriter(parameters, analysis), inPredicate);
-        translationMap.addIntermediateMapping(inPredicate, parametersReplaced);
         SymbolReference valueList = getOnlyElement(subquery.getOutputSymbols()).toSymbolReference();
-        translationMap.addIntermediateMapping(parametersReplaced, new InPredicate(parametersReplaced.getValue(), valueList));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: does `addIntermediateMapping` can be removed? It would be nice to make TranslationMap to be immutable.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java]: I just meant the method extraction, not changes in it

**Comment by kokosing** [comment]
@martint Can you please take a look? This commit much simplifies rewrites of `ApplyNode`s! This is prerequisite to continue work on other subquery related work like fixing coercions and lateral join.

**Comment by sopel39** [comment]
superceded by: https://github.com/prestodb/presto/pull/6727
